File: appkit-creating-and-customizing-the-touch-bar/Objective-C/NSTouchBar Catalog/TestViewControllers/ButtonViewController.m
/*
See the LICENSE.txt file for this sample’s licensing information.

Abstract:
View controller responsible for showing NSStepperTouchBarItem in an NSTouchBar instance.
*/

#import "ButtonViewController.h"

static NSTouchBarCustomizationIdentifier ButtonItemCustomizationIdentifier = @"com.TouchBarCatalog.buttonItemController";
static NSTouchBarItemIdentifier ButtonItemIdentifier = @"com.TouchBarCatalog.buttonItem";

@interface ButtonViewController () <NSTouchBarDelegate>

@property (strong) NSButtonTouchBarItem *buttonTouchBarItem;

@property (weak) IBOutlet NSButton *useCustomColor;

@end


#pragma mark -

@implementation ButtonViewController

// Note: This particular view controller doesn't allow customizing its NSTouchBar instance.

- (void)viewDidLoad
{
    [super viewDidLoad];
}

// The user chose Custom Button Color checkbox to change the button colors.
- (IBAction)customize:(id)sender
{
    // Change each button to the right background color.
    for (NSTouchBarItemIdentifier itemIdentifier in self.touchBar.itemIdentifiers)
    {
        NSButtonTouchBarItem *touchBarItem = [self.touchBar itemForIdentifier:itemIdentifier];
        NSButton *button = (NSButton *)touchBarItem.view;
        button.bezelColor = (self.useCustomColor.state == NSControlStateValueOn) ? [NSColor systemYellowColor] : nil;
    }
}

- (void)buttonAction:(id)sender
//••- (IBAction)buttonAction:(id)sender
{
    NSLog(@"%@ was pressed", ((NSButtonTouchBarItem *)sender).title);
}

#pragma mark NSTouchBarProvider

- (NSTouchBar *)makeTouchBar
{
    NSTouchBar *bar = [[NSTouchBar alloc] init];
    bar.delegate = self;
    
    bar.customizationIdentifier = ButtonItemCustomizationIdentifier;
    
    // Set the default ordering of items.
    bar.defaultItemIdentifiers =
        @[ButtonItemIdentifier, NSTouchBarItemIdentifierOtherItemsProxy];
    
    bar.customizationAllowedItemIdentifiers = @[ButtonItemIdentifier];
    
    return bar;
}

#pragma mark NSTouchBarDelegate

// The system calls this while constructing the NSTouchBar for each NSTouchBarItem you want to create.
- (nullable NSTouchBarItem *)touchBar:(NSTouchBar *)touchBar makeItemForIdentifier:(NSTouchBarItemIdentifier)identifier
{
    if ([identifier isEqualToString:ButtonItemIdentifier])
    {
        _buttonTouchBarItem = [[NSButtonTouchBarItem alloc] initWithIdentifier:ButtonItemIdentifier];
        
        self.buttonTouchBarItem.title = NSLocalizedString(@"Button", @"");
        self.buttonTouchBarItem.target = self;
        self.buttonTouchBarItem.action = @selector(buttonAction:);
        self.buttonTouchBarItem.image = [NSImage imageWithSystemSymbolName:@"tray" accessibilityDescription:@"tray"];
        
        return self.buttonTouchBarItem;
    }
    
    return nil;
}

@end

File: appkit-creating-and-customizing-the-touch-bar/Swift/NSTouchBar Catalog/ButtonViewController.swift
/*
See the LICENSE.txt file for this sample’s licensing information.

Abstract:
View controller responsible for showing NSButtonTouchBarItem in an NSTouchBar instance.
*/

import Cocoa

class ButtonViewController: NSViewController {
    let button = NSTouchBarItem.Identifier("com.TouchBarCatalog.TouchBarItem.button")
    let buttonBar = NSTouchBar.CustomizationIdentifier("com.TouchBarCatalog.buttonBar")
    
    @IBOutlet weak var sizeConstraint: NSButton!
    @IBOutlet weak var useCustomColor: NSButton!
    
    // MARK: - Action Functions
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    override func makeTouchBar() -> NSTouchBar? {
        let touchBar = NSTouchBar()
        touchBar.delegate = self
        touchBar.customizationIdentifier = buttonBar
        touchBar.defaultItemIdentifiers = [button]
        return touchBar
    }
    
    @IBAction func customize(_ sender: AnyObject) {
        guard let touchBar = touchBar else { return }
        
        for itemIdentifier in touchBar.itemIdentifiers {
            guard let item = touchBar.item(forIdentifier: itemIdentifier) as? NSButtonTouchBarItem,
                let button = item.view as? NSButton else { continue }

            button.bezelColor = useCustomColor.state == NSControl.StateValue.on ? NSColor.systemYellow : nil
        }
    }
    
    @IBAction func buttonAction(_ sender: AnyObject) {
        if let button = sender as? NSButtonTouchBarItem {
            print("\(#function): button with title \"\(button.title)\" is tapped")
        }
    }
    
}

// MARK: - NSTouchBarDelegate

extension ButtonViewController: NSTouchBarDelegate {

    // The system calls this while constructing the NSTouchBar for each NSTouchBarItem you want to create.
    func touchBar(_ touchBar: NSTouchBar, makeItemForIdentifier identifier: NSTouchBarItem.Identifier) -> NSTouchBarItem? {
        switch identifier {
        case button:
            let buttonItem = NSButtonTouchBarItem(identifier: identifier)
            buttonItem.title = NSLocalizedString("Button", comment: "")
            buttonItem.target = self
            buttonItem.action = #selector(buttonAction)
            buttonItem.image = NSImage(systemSymbolName: "tray", accessibilityDescription: "tray")
            return buttonItem
        default:
            return nil
        }
    }

}
